CBinding/Compiler/ICompiler.cs

```diff
@@ -30,47 +30,431 @@
 // WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 //
 
+using System;
+using System.Threading.Tasks;
+using System.IO;
+using System.Text;
+using System.Linq;
+
 using Mono.Addins;
 
 using MonoDevelop.Core;
 using MonoDevelop.Projects;
+using MonoDevelop.Core.Execution;
+
 
 namespace CBinding
 {
-	[TypeExtensionPoint ("/CBinding/Compilers")]
 	public interface ICompiler
 	{
 		string Name {
 			get;
 		}
-		
+
 		Language Language {
 			get;
 		}
-		
+
 		string CompilerCommand {
 			get;
 		}
-		
+
 		bool SupportsCcache {
 			get;
 		}
-		
+
 		bool SupportsPrecompiledHeaders {
 			get;
 		}
-		
+
 		string GetCompilerFlags (Project project, CProjectConfiguration configuration);
-		
+
 		string GetDefineFlags (Project project, CProjectConfiguration configuration);
-		
+
 		BuildResult Compile (
-		    Project project,
-		    ProjectFileCollection projectFiles,
-		    ProjectPackageCollection packages,
-		    CProjectConfiguration configuration,
-		    ProgressMonitor monitor);
-		
+			Project project,
+			ProjectFileCollection projectFiles,
+			ProjectPackageCollection packages,
+			CProjectConfiguration configuration,
+			ProgressMonitor monitor);
+
 		void Clean (ProjectFileCollection projectFiles, CProjectConfiguration configuration, ProgressMonitor monitor);
 	}
-}
+
+	[TypeExtensionPoint("/CBinding/Toolchains")]
+	public abstract class CMakeToolchain : FolderBasedProject
+	{
+		FilePath file;
+		public override FilePath FileName {
+			get {
+				return file;
+			}
+			set {
+				file = value;
+			}
+		}
+
+		/// The name of this toolchain.
+		public abstract string ToolchainName { 
```

Comment:
I read up on this a little more. Apparently the -G argument is used for "generators", and CMake has a different concept called "toolchains". 

https://cmake.org/cmake/help/v3.0/manual/cmake-generators.7.html

Maybe we should rename the class from CMakeToolchain to CMakeGenerator.

I'd probably do something like have CMakeGenerator be a concrete class, not abstract. It would then do something like:

```csharp
string arguments = "../";
if (!string.IsNullOrEmpty(ID)) {
  arguments += $" -G \"{ID}\"";  
}
Stream generationResult = ExecuteCommand ("cmake", args, outputDirectory, monitor);
```

and its `ID` would be empty. The subclasses would override it and provide ID values.